enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,435 @@
+
+**Note**: This phase may require manual upgrades of OLM-based operators installed through this PO mechanism. The process for manually upgrading PO packages will be detailed below.
+
+In addition, future phase requirements around package dependency resolution may require a new design entirely. While this is a notable short term trade-off, it allows the OLM team to continue iterating on their "deppy" component in the background. Deppy is an OLM 1.x component that's responsible for dependency resolution. A future design may elect to use deppy and it's APIs, and delegate package selection to that component. The usage of feature flags can be used to toggle the underlying implementation layer once that component is ready.
+
+This phase will likely restrict which OLM-based operators admins can install. Those requirements will be detailed below. The immediate pool of candidates may include the cert-manager and service-binding-operator operators.a
+
+#### Phase 1: Expand Capabilities
+
+Summary: Introduce depedency resolution, and expand the set of supported capabilities for admins and platform operator authors. Help establish, or improve, any first-class APIs at the installer and cluster runtime levels.
+
+The previous phase aimed to get the foundational layer in place to start onboarding more OLM-based operators to participate in the cluster's lifecycle. The notable limitations with phase 0 is requiring manual admin intervention around editing the installer's Kubernetes manifests and upgrading PO packages after installation. This phase will aim to address those limitations by introducing first-class concepts.
+
+This phase will also attempt to introduce package management and dependency resolution primitives. Candidate PO packages may place dependencies on other packages and GVKs, in addition to more granular constraints that are needed in order to accommodate EUS-EUS drift and platform-specific cluster invariants.
+
+#### Phase 2: QoL Improvements
+
+Summary: Prepare for general availability by tackling any performance improvements, shoring up the overall design/implementation/UX, and any QoL improvements that have been identified from previous phase feedback.
+
+This phase has largely been unexplored. Candidates for this phase may include:
+
+- Integration with the OCP console, hypershift, KCP, etc.
+- Improved instrumentation.
+- Improved performance of the overall stack.
+- Improved UX around the user-facing APIs.
+
+### PO Requirements
+
+- Must package Kubernetes manifests using OLM's registry+v1 bundle format.
+- Cannot specify non-AllNamespace CSV install modes.
+- Cannot declare package or GVK dependencies.
+- All package bundles must live in the redhat-operators CatalogSource.
```

Comment:
I think where I'm a bit hazy is how much control we have over catalog content. My preference would be to require that all POs are sourced from a dedicated index image, and that index image only contains supported PO packages that all belong to the same channel but I wasn't sure how realistic this preference was given there's potentially external team involvement here.

But to answer you question, we could make this a requirement but it depends on much flexibility we have here.

Reply:
i think the two things are likely orthogonal-ish.

It may be true that to be a PO, you need to confirm to a particular channel name convention that we define.

But that won't be sufficient to identify POs

Honestly it might just be metadata on the bundle that annotates it as a PO.

Reply:
i'm curious what this requirement might mean for platform integrators that wish to provide infrastructure specific operators, would they need to have some blessing from red hat to use the redhat-operators CatalogSource?

Reply:
I think the redhat-operators CatalogSource is just a short term design trade-off that's scoped to phase 0, and we'll see iron out these requirements over the next phase or two. There's no immediate technical limitation here around sourcing PO packages from any catalog source in the cluster, but as Ben alluded to in a previous comment, we're lacking the bundle-level identifier in the short term, and this may result in non-determinism when finding the desired set of packages.

Reply:
We will definitely need to give partners the ability to create and deliver platform operators independently of Red Hat, but that can come in a later phase.

Reply:
I think Ben suggested this, but what if (for now) we have the PO controller consume all of the CatalogSources in the cluster, and then a separate PO-specific configmap that contains a list of <catalog source, package, channel> tuples that tell the PO controller which operators to consider.

That would a) give us the ability to ship a default config with a handful of known good operators folks can play with and b) give operator authors the ability to exercise their operator as a PO without requiring our intervention.

> Honestly it might just be metadata on the bundle that annotates it as a PO.

I'd prefer to say metadata on a channel. If its on a bundle, that opens the door to possibilities of some bundles in the channel having the annotation and some bundles not, which makes the PO upgrade story a bit precarious.

Reply:
> I'd prefer to say metadata on a channel.

yeah, fair.

Reply:
Added this as an answered question in the latest commit. Hinted towards defining channel-level properties for identifying POs long term, and mentioned the ConfigMap implementation that we had discussed in slack last week as a potential nice-to-have but not blocking for the initial design.